packages/python/plotly/plotly/express/_core.py

```diff
@@ -1648,109 +1779,193 @@ def process_dataframe_hierarchy(args):
+                )
+            else:
+                df = df.with_columns(
+                    **{args["values"]: nw.col(args["values"]).cast(nw.Float64())}
+                )
+
+        except Exception:  # pandas, Polars and pyarrow exception types are different
             raise ValueError(
                 "Column `%s` of `df` could not be converted to a numerical data type."
                 % args["values"]
             )
 
-        if args["color"]:
-            if args["color"] == args["values"]:
-                new_value_col_name = args["values"] + "_sum"
-                df[new_value_col_name] = df[args["values"]]
-                args["values"] = new_value_col_name
+        if args["color"] and args["color"] == args["values"]:
+            new_value_col_name = args["values"] + "_sum"
+            df = df.with_columns(**{new_value_col_name: nw.col(args["values"])})
+            args["values"] = new_value_col_name
         count_colname = args["values"]
     else:
         # we need a count column for the first groupby and the weighted mean of color
         # trick to be sure the col name is unused: take the sum of existing names
+        columns = df.collect_schema().names()
         count_colname = (
-            "count"
-            if "count" not in df.columns
-            else "".join([str(el) for el in list(df.columns)])
+            "count" if "count" not in columns else "".join([str(el) for el in columns])
         )
         # we can modify df because it's a copy of the px argument
-        df[count_colname] = 1
+        df = df.with_columns(**{count_colname: nw.lit(1)})
         args["values"] = count_colname
-    agg_f[count_colname] = "sum"
+
+    # Since count_colname is always in agg_f, it can be used later to normalize color
+    # in the continuous case after some gymnastic
+    agg_f[count_colname] = nw.sum(count_colname)
+
+    discrete_aggs = []
+    continuous_aggs = []
 
     if args["color"]:
-        if not _is_continuous(df, args["color"]):
-            aggfunc_color = aggfunc_discrete
-            discrete_color = True
+        if discrete_color:
+
+            discrete_aggs.append(args["color"])
+            # Hack: In theory, we should have a way to do `.agg(nw.col(x).unique())` and
+            # successively unpack/parse it as:
+            # ```
+            # (nw.when(nw.col(x).list.len()==1)
+            # .then(nw.col(x).list.first())
+            # .otherwise(nw.lit("(?)"))
+            # )
+            # ```
+            # which replicates:
+            # ```
+            # def discrete_agg(x):
+            #     uniques = x.unique()
+            #     return uniques[0] if len(uniques) == 1 else "(?)"
+            # ```
+            # However we cannot do that just yet, therefore a workaround is provided
+            agg_f[args["color"]] = nw.col(args["color"]).max()
+            agg_f[f'{args["color"]}__plotly_n_unique__'] = (
```

Comment:
should we use `generate_unique_token` here?

Reply:
Just to make sure I get this right, do you mean `f'{args["color"]}' + <token>` or just `<token>`?
The former case would keep similar comfort in the workaround but there is no guarantee of uniqueness (since it compares with columns that do not contain `f'{args["color"]}' + <token>`. The latter case would be a fairly additional headache since we do that repeatedly the same workaround in line 1882, and we should keep track of all the tokens added so far, and to which column they refer to (since they come in pairs).

We can go really specific in the suffix info (e.g. `__plotly_process_dataframe_hierarchy_discrete_agg_n_unique__`), hopefully no one uses a suffix like this one 😁

Reply:
We can use `f'{args["color"]}' + <token>`, and use some long token (16 bytes?), I hope probability of having a same random token + column name shrinks fast